presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java

```diff
@@ -114,27 +143,32 @@ public List<String> listSchemaNames(ConnectorSession session)
     @Override
     public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
     {
-        IcebergTableName name = IcebergTableName.from(tableName.getTableName());
-        verify(name.getTableType() == DATA, "Wrong table type: " + name.getTableType());
+        try {
+            IcebergTableName name = IcebergTableName.from(tableName.getTableName());
+            verify(name.getTableType() == DATA, "Wrong table type: " + name.getTableType());
+
+            MetastoreContext metastoreContext = new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), Optional.empty(), false, HiveColumnConverterProvider.DEFAULT_COLUMN_CONVERTER_PROVIDER);
+            Optional<Table> hiveTable = metastore.getTable(metastoreContext, tableName.getSchemaName(), name.getTableName());
+            if (!hiveTable.isPresent()) {
+                return null;
+            }
+            if (!isIcebergTable(hiveTable.get())) {
+                throw new UnknownTableTypeException(tableName);
+            }
 
-        MetastoreContext metastoreContext = new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), Optional.empty(), false, HiveColumnConverterProvider.DEFAULT_COLUMN_CONVERTER_PROVIDER);
-        Optional<Table> hiveTable = metastore.getTable(metastoreContext, tableName.getSchemaName(), name.getTableName());
-        if (!hiveTable.isPresent()) {
-            return null;
+            org.apache.iceberg.Table table = getHiveIcebergTable(metastore, hdfsEnvironment, session, tableName);
+            Optional<Long> snapshotId = getSnapshotId(table, name.getSnapshotId());
+
+            return new IcebergTableHandle(
+                    tableName.getSchemaName(),
+                    name.getTableName(),
+                    name.getTableType(),
+                    snapshotId,
+                    TupleDomain.all());
         }
-        if (!isIcebergTable(hiveTable.get())) {
-            throw new UnknownTableTypeException(tableName);
+        catch (TableNotFoundException ex) {
```

Comment:
Let me revert back on this. Unable to recall, will have to debug.

Reply:
getSnapshotId throws this. I am pasting the stacktrace for reference.

Reply:
```
[ERROR] Tests run: 1657, Failures: 1, Errors: 0, Skipped: 17, Time elapsed: 1,619.726 s <<< FAILURE! - in TestSuite
2023-09-01T15:21:51.9988108Z [ERROR] com.facebook.presto.iceberg.hive.TestIcebergDistributedHive.testViewMetadata  Time elapsed: 0.598 s  <<< FAILURE!
2023-09-01T15:21:51.9990516Z java.lang.RuntimeException: Table 'tpch.meta_test_view' not found
2023-09-01T15:21:51.9991487Z 	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
2023-09-01T15:21:51.9992457Z 	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:733)
2023-09-01T15:21:51.9993493Z 	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:701)
2023-09-01T15:21:51.9994448Z 	at com.facebook.presto.tests.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:132)
2023-09-01T15:21:51.9995519Z 	at com.facebook.presto.tests.AbstractTestQueryFramework.computeActual(AbstractTestQueryFramework.java:127)
2023-09-01T15:21:51.9996577Z 	at com.facebook.presto.tests.AbstractTestDistributedQueries.testViewMetadata(AbstractTestDistributedQueries.java:941)
2023-09-01T15:21:51.9997486Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-09-01T15:21:51.9998202Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2023-09-01T15:21:51.9999176Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-09-01T15:21:51.9999879Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2023-09-01T15:21:52.0000661Z 	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
2023-09-01T15:21:52.0001540Z 	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
2023-09-01T15:21:52.0002465Z 	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
2023-09-01T15:21:52.0003281Z 	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
2023-09-01T15:21:52.0004085Z 	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
2023-09-01T15:21:52.0004899Z 	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
2023-09-01T15:21:52.0006168Z 	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
2023-09-01T15:21:52.0006996Z 	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
2023-09-01T15:21:52.0008315Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2023-09-01T15:21:52.0009257Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2023-09-01T15:21:52.0009864Z 	at java.lang.Thread.run(Thread.java:750)
2023-09-01T15:21:52.0011196Z Caused by: com.facebook.presto.spi.TableNotFoundException: Table 'tpch.meta_test_view' not found
2023-09-01T15:21:52.0012076Z 	at com.facebook.presto.iceberg.HiveTableOperations.refresh(HiveTableOperations.java:211)
2023-09-01T15:21:52.0013227Z 	at com.facebook.presto.iceberg.HiveTableOperations.current(HiveTableOperations.java:191)
2023-09-01T15:21:52.0014000Z 	at org.apache.iceberg.BaseTable.currentSnapshot(BaseTable.java:130)
2023-09-01T15:21:52.0015237Z 	at com.facebook.presto.iceberg.IcebergHiveMetadata.getSnapshotId(IcebergHiveMetadata.java:511)
2023-09-01T15:21:52.0016382Z 	at com.facebook.presto.iceberg.IcebergHiveMetadata.getTableHandle(IcebergHiveMetadata.java:148)
2023-09-01T15:21:52.0017534Z 	at com.facebook.presto.iceberg.IcebergHiveMetadata.getTableHandle(IcebergHiveMetadata.java:105)
2023-09-01T15:21:52.0020417Z 	at com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.getTableHandle(ClassLoaderSafeConnectorMetadata.java:220)
2023-09-01T15:21:52.0021616Z 	at com.facebook.presto.metadata.MetadataUtil.getOptionalTableHandle(MetadataUtil.java:180)
2023-09-01T15:21:52.0022644Z 	at com.facebook.presto.metadata.MetadataManager$1.getTableHandle(MetadataManager.java:1331)
2023-09-01T15:21:52.0024774Z 	at com.facebook.presto.connector.informationSchema.InformationSchemaMetadata.lambda$calculatePrefixesWithTableName$6(InformationSchemaMetadata.java:299)
2023-09-01T15:21:52.0025806Z 	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
2023-09-01T15:21:52.0026537Z 	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
2023-09-01T15:21:52.0027214Z 	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
2023-09-01T15:21:52.0027906Z 	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
2023-09-01T15:21:52.0028563Z 	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
2023-09-01T15:21:52.0029249Z 	at java.util.Iterator.forEachRemaining(Iterator.java:116)
2023-09-01T15:21:52.0029912Z 	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
2023-09-01T15:21:52.0030649Z 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
2023-09-01T15:21:52.0031382Z 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
2023-09-01T15:21:52.0032129Z 	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
2023-09-01T15:21:52.0032843Z 	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
2023-09-01T15:21:52.0034172Z 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2023-09-01T15:21:52.0035021Z 	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
2023-09-01T15:21:52.0035732Z 	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:272)
2023-09-01T15:21:52.0036373Z 	at java.util.Iterator.forEachRemaining(Iterator.java:116)
2023-09-01T15:21:52.0037117Z 	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
2023-09-01T15:21:52.0037971Z 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
2023-09-01T15:21:52.0038962Z 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
2023-09-01T15:21:52.0039877Z 	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
2023-09-01T15:21:52.0040794Z 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
2023-09-01T15:21:52.0041754Z 	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
2023-09-01T15:21:52.0044292Z 	at com.facebook.presto.connector.informationSchema.InformationSchemaMetadata.calculatePrefixesWithTableName(InformationSchemaMetadata.java:301)
2023-09-01T15:21:52.0045744Z 	at com.facebook.presto.connector.informationSchema.InformationSchemaMetadata.getTableLayouts(InformationSchemaMetadata.java:243)
2023-09-01T15:21:52.0046816Z 	at com.facebook.presto.metadata.MetadataManager.getLayout(MetadataManager.java:360)
2023-09-01T15:21:52.0047887Z 	at com.facebook.presto.sql.planner.iterative.rule.PickTableLayout.pushPredicateIntoTableScan(PickTableLayout.java:307)
2023-09-01T15:21:52.0049285Z 	at com.facebook.presto.sql.planner.iterative.rule.PickTableLayout.pushPredicateIntoTableScan(PickTableLayout.java:252)
2023-09-01T15:21:52.0051029Z 	at com.facebook.presto.sql.planner.iterative.rule.PickTableLayout$PickTableLayoutForPredicate.apply(PickTableLayout.java:137)
2023-09-01T15:21:52.0052531Z 	at com.facebook.presto.sql.planner.iterative.rule.PickTableLayout$PickTableLayoutForPredicate.apply(PickTableLayout.java:102)
2023-09-01T15:21:52.0053681Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.transform(IterativeOptimizer.java:223)
2023-09-01T15:21:52.0055222Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreNode(IterativeOptimizer.java:184)
2023-09-01T15:21:52.0056271Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:146)
2023-09-01T15:21:52.0057509Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:267)
2023-09-01T15:21:52.0058586Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:148)
2023-09-01T15:21:52.0060079Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:267)
2023-09-01T15:21:52.0061297Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:148)
2023-09-01T15:21:52.0062376Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:267)
2023-09-01T15:21:52.0063444Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:148)
2023-09-01T15:21:52.0064698Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:267)
2023-09-01T15:21:52.0065939Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:148)
2023-09-01T15:21:52.0067102Z 	at com.facebook.presto.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:133)
2023-09-01T15:21:52.0068013Z 	at com.facebook.presto.sql.Optimizer.validateAndOptimizePlan(Optimizer.java:116)
2023-09-01T15:21:52.0069586Z 	at com.facebook.presto.execution.SqlQueryExecution.lambda$createLogicalPlanAndOptimize$3(SqlQueryExecution.java:544)
2023-09-01T15:21:52.0070477Z 	at com.facebook.presto.common.RuntimeStats.profileNanos(RuntimeStats.java:136)
2023-09-01T15:21:52.0072373Z 	at com.facebook.presto.execution.SqlQueryExecution.createLogicalPlanAndOptimize(SqlQueryExecution.java:542)
2023-09-01T15:21:52.0075166Z 	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:451)
2023-09-01T15:21:52.0082081Z 	at com.facebook.presto.$gen.Presto_0_284_SNAPSHOT_58ba3bd__testversion____20230901_151309_17207.run(Unknown Source)
2023-09-01T15:21:52.0082816Z 	at com.facebook.presto.execution.SqlQueryManager.createQuery(SqlQueryManager.java:306)
2023-09-01T15:21:52.0083928Z 	at com.facebook.presto.dispatcher.LocalDispatchQuery.lambda$startExecution$8(LocalDispatchQuery.java:211)
2023-09-01T15:21:52.0084443Z 	... 3 more
```

Reply:
@pratyakshsharma Thanks for the stack. Was the view tpch.meta_test_view actually not exist when this exception happens? I don't see Trino has to catch this exception in getTableHandle(). Can you run the same query or test on Trino to see how they handled it?

Reply:
The view exists, but since at this point presto is trying to identify a "table" with the same name, this exception is thrown. Sure let me check and get back.

Reply:
@pratyakshsharma Any update on this?

Reply:
@yingsu00 this test does not exist on trino side. I tried to debug the show columns flow in trino to see how they are handling it. Basically when we read information_schema.columns table in presto, there is a existing check to see if a view or table exists with the qualified object name. This check is present in `InformationSchemaMetadata. calculatePrefixesWithTableName` method. 
Trino however has removed this check from this method as part of their redirection changes in iceberg connector. They now perform these checks at a later stage in the flow as mentioned in this PR - https://github.com/trinodb/trino/pull/18730

Reply:
> @yingsu00 this test does not exist on trino side. I tried to debug the show columns flow in trino to see how they are handling it. Basically when we read information_schema.columns table in presto, there is a existing check to see if a view or table exists with the qualified object name, and it gets called from IterativeOptimizer. This check is present in `InformationSchemaMetadata. calculatePrefixesWithTableName` method. Trino however has removed this check from this method as part of their redirection changes in iceberg connector. They now perform these checks at a later stage in the flow as mentioned in this PR - [trinodb/trino#18730](https://github.com/trinodb/trino/pull/18730)

@pratyakshsharma What test is this?

Reply:
@pratyakshsharma I took a look at the code and test. This TestIcebergDistributedHive.testViewMetadata test runs this query:
```
CREATE VIEW meta_test_view AS SELECT BIGINT '123' x, 'foo' y;
SHOW COLUMNS FROM meta_test_view;
```
The query `SHOW COLUMNS FROM meta_test_view` reads the `information_schema.columns` table and filter out all rows that the `table_name` column is not "meta_test_view". To do this, it needs to apply the constraint of two domains (table_schema = 'tpch', table_name='meta_test_view') on `information_schema.columns` table. This happens filtering satisfying prefixes in `calculatePrefixesWithTableName()`:
```
if (tables.isPresent()) {
            return prefixes.stream()
                    .flatMap(prefix -> tables.get().stream()
                            .filter(this::isLowerCase)
                            .map(table -> table.toLowerCase(ENGLISH))
                            .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
                    .filter(objectName -> metadata.getMetadataResolver(session).getTableHandle(objectName).isPresent() || metadataResolver.getView(objectName).isPresent())
                    .map(QualifiedTablePrefix::toQualifiedTablePrefix)
                    .collect(toImmutableSet());
        }
```
Here metadata.getMetadataResolver(session).getTableHandle(objectName) throws exception because meta_test_view is a view, not an Iceberg table and doesn't have a snapshot. That's why you added the try-catch in getTableHandle(). But in fact, we should move the view matching BEFORE matching the tableHandle. This is what Trino does as well. Here is their `calculatePrefixesWithTableName()` implementation. Note my comment below:
```
Set<QualifiedTablePrefix> tablePrefixes = prefixes.stream()
                    .peek(prefix -> verify(prefix.asQualifiedObjectName().isEmpty()))
                    .flatMap(prefix -> prefix.getSchemaName()
                            .map(schemaName -> Stream.of(prefix))
                            .orElseGet(() -> listSchemaNames(session)))
                    .flatMap(prefix -> tables.get().stream()
                            .filter(this::isLowerCase)
                            .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
                    .filter(objectName -> {
                        if (!isColumnsEnumeratingTable(informationSchemaTable) ||
                                metadata.isMaterializedView(session, objectName) ||
                                metadata.isView(session, objectName)) {
                            return true;    // It returns here because it's a view. It does NOT need to redirect the tables at all.
                        }

                        // This is a columns enumerating table and the object is not a view
                        try {
                            // Table redirection to enumerate columns from target table happens later in
                            // MetadataListing#listTableColumns, but also applying it here to avoid incorrect
                            // filtering in case the source table does not exist or there is a problem with redirection.
                            return metadata.getRedirectionAwareTableHandle(session, objectName).getTableHandle().isPresent();
                        }
                        catch (TrinoException e) {
                            if (e.getErrorCode().equals(TABLE_REDIRECTION_ERROR.toErrorCode())) {
                                // Ignore redirection errors for listing, treat as if the table does not exist
                                return false;
                            }

                            throw e;
                        }
                    })
                    .filter(objectName -> predicate.isEmpty() || predicate.get().test(asFixedValues(objectName)))
                    .map(QualifiedObjectName::asQualifiedTablePrefix)
                    .distinct()
                    .limit(MAX_PREFIXES_COUNT + 1)
                    .collect(toImmutableSet());
```
Can you do the same to see if it works?

Reply:
Shifting view check before table actually helped. I was unnecessarily making it complicated.